src/theory/quantifiers/fun_def_process.cpp

```diff
@@ -169,6 +179,29 @@ Node FunDefFmf::simplifyFormula( Node n, bool pol, bool hasPol, std::vector< Nod
         if( childChanged ){
           nn = NodeManager::currentNM()->mkNode( n.getKind(), children );
         }
+        if( branch_pos && !constraints.empty() ){
+          // if we are at a branching position in the formula, we can
+          // minimize recursive constraints on recursively defined predicates if we know one child forces
+          // the overall evaluation of this formula.
+          Node branch_cond;
+          if( n.getKind()==ITE ){
+            // always care about constraints on the head of the ITE, but only care about one of the children depending on how it evaluates
+            branch_cond = NodeManager::currentNM()->mkNode( kind::AND, branch_constraints[0],
+                            NodeManager::currentNM()->mkNode( kind::ITE, n[0], branch_constraints[1], branch_constraints[2] ) );
+          }else{
+            // in the default case, we care about all conditions
+            branch_cond = constraints.size()==1 ? constraints[0] : NodeManager::currentNM()->mkNode( AND, constraints );
+            for( unsigned i=0; i<n.getNumChildren(); i++ ){
+              // if this child holds with forcing polarity (true child of OR or false child of AND), 
+              // then we only care about its associated recursive conditions 
+              branch_cond = NodeManager::currentNM()->mkNode( kind::ITE, 
+                              ( n.getKind()==OR ? n[i] : n[i].negate() ), branch_constraints[i], branch_cond );
+            }
+          }
+          Trace("fmf-fun-def-debug2") << "Made branching condition " << branch_cond << std::endl;
+          constraints.clear();
+          constraints.push_back( branch_cond );
```

Comment:
I find "constraints" fairly confusing in this function. In particular, what is its contract? Can you please the contract in the header.

Reply:
"constraints" is the second part of the pair returned by the function A_0 in Figure 1 of http://homepage.divms.uiowa.edu/~ajreynol/ijcar16a.pdf.

I've added comments in the *.h to explain this.

Reply:
We should probably consider making it a part of the return value in the future then.